### ISSUE TITLE
Packet handler and view plugins for 0.75

### ISSUE THREAD

**Issue by sven-n** [MUnique/OpenMU#111]
The version 0.75 is very old and contains just a subset of the features of this server.
Some packet structures might not have been changed from 0.75 to season 6, so there is probably not much work to do. From a historical point of view it’s also interesting how messages passed between client and server got changed or extended over time. The simplicity of the early server also might reveal missing but important messages in season 6.

Another thing to note: If a message structure of season 6 from server to client contains additional fields (but rest is valid), it’s not required to remove these fields for 0.75. The client would just not map these fields to a struct field.

I analyzed what needs to be changed. The following list can be seen as a to do list, too. Italic entries were supported initially.

General:
  * [x] Appearance Serializer
  * [x] Item Serializer

Packet Handlers:
  * [x]  0xF1 - Login Group
    * [x] 0x01 - Login Request
    * [x] *0x02 - Log out Request*
  * [x] *0xF3 - Character Group*
    * [x] *0x00 - Request Character List*
    * [x] *0x01 - Character Create Request*
    * [x] *0x02 - Character Delete Request*
    * [x] *0x03 - Character Selection Request*
    * [x] *0x06 - Add level up point*
    * [x] *0x12 - Data loaded*
  * [x] *0x00 - Chat message*
  * [x] *0x01 - Whisper message*
  * [x] 0x10 - Move/walk
  * [x] 0x11 - Set position (e.g. by skills)
  * [x] 0x15 - Hit
  * [x] *0x18 - Animation*
  * [x] 0x19 - Skill
  * [x] 0x1C - Teleport
  * [x] 0x1D - AreaSkillHit
  * [x] 0x1E - Area Skill
  * [x] *0x22 - Item pickup request*
  * [x] *0x23 - Item drop request*
  * [x] *0x24 - Item move request*
  * [x] *0x26 - Item use/consume request*
  * [x] *0x30 - Talk npc request*
  * [x] *0x32 - Item buy request*
  * [x] *0x33 - Item sell request*
  * [x] *0x36 - Trade request*
  * [x] *0x37 - Trade request response*
  * [x] *0x3A - Trade money*
  * [x] *0x3C - Trade button state change*
  * [x] *0x3D - Trade finished*
  * [x] *0x40 - Party request*
  * [x] *0x41 - Party request response*
  * [x] *0x42 - Party list request*
  * [x] *0x43 - Party member kick*
  * [x] *0x50 - Guild request*
  * [x] *0x51 - Guild request response*
  * [x] *0x52 - Guild list request*
  * [x] *0x53 - Guild member kick*
  * [x] *0x54 - Guild master answer*
  * [x] *0x55 - Guild master info save*
  * [x] 0x57 - Guild master cancel creation
  * [x] 0x61 - Request guild war
  * [x] *0x81 - Vault money in/out*
  * [x] *0x82 - Vault close*
  * [x] *0x86 - Chaos machine mix*
  * [x] *0x87 - Chaos machine close*
  
View plugins:
  * [x] *0xF1 - Login Group*
    * [x] *0x01 - Login result*
    * [x] *0x02 - Logout*
  * [x] 0xF3 - Character Group
    * [x] 0x00 - Character list
    * [x] 0x01 - Character create result
    * [x] *0x02 - Character delete result*
    * [x] 0x03 - Character stats
    * [x] 0x04 - Character respawn after death
    * [x] *0x05 - Character level up*
    * [x] *0x06 - Character stat increase result*
    * [x] 0x07 - Character damage (poison)
    * [x] *0x08 - Player kill level changed*
    * [x] *0x10 - Inventory*
    * [x] 0x11 - Skill list
    * [ ] 0x13 - Player equipment
      * sent after cancelled trade, required?
    * [ ] 0x14 - Player item modified
      * sent after item usage/consumption, required?
    * [x] 0x20 - Guild soccer life?
    * [x] 0x22 - Guild soccer time update
    * [x] 0x23 - Guild soccer goal
  * [x] 0x0F - Weather update
  * [x] 0x10 - Object walked
  * [x] 0x11 - Object position update
  * [x] 0x12 - Player meet
    * Required changes: 0.75 doesn't send the active effects at the end of each player block; Additional field "ViewSkillState" (poisoned, iced) is required.
  * [x] 0x13 - NPC/Monster in scope
    * Required changes: 0.75 doesn't send the active effects at the end of each npc block; Additional field "ViewState" (poisoned, iced) is required; npc type is 1 byte instead of 2 bytes.
  * [x] 0x1F - Summoned monster in scope
  * [x] *0x14 - Object out of scope*
  * [x] 0x15 - Attack
  * [x] *0x16 - Experience gained by killed monster*
  * [x] *0x17 - Player died*
  * [x] *0x18 - Animation*
  * [x] 0x19 - Skill
  * [x] 0x1B - Skill cancelled (e.g. disable poison/ice effect)
  * [x] 0x1C - Teleport result
  * [x] 0x1D - Area skill hit response - is it even required?!
  * [x] 0x1E - Area Skill
  * [x] *0x20 - Item dropped*
  * [x] *0x22 - Item pick up result*
  * [x] *0x23 - Item drop result*
  * [x] *0x24 - Item move result*
  * [x] *0x25 - Player equipment changed*
  * [x] *0x26 - Health state update*
  * [x] *0x27 - Mana state update*
  * [x] *0x28 - Item consumed/removed*
  * [x] *0x29 - Special item usage (Ale)*
  * [x] *0x2A - Item durability change*
  * [x] *0x30 - npc talk result*
  * [x] *0x31 - npc shop items*
  * [x] *0x32 - npc item buy result*
  * [x] *0x33 - npc item sell result*
  * [x] *0x36 - trade request*
  * [x] *0x37 - trade request response*
  * [x] *0x38 - trade item removed*
  * [x] *0x39 - trade item added*
  * [x] *0x3A - trade money changed (own player)*
  * [x] *0x3B - trade money changed (other player)*
  * [x] *0x3C - Trade button state changed*
  * [x] *0x3D - trade finished*
  * [x] *0x40 - party request*
  * [x] 0x42 - party list
    * 0.75 doesn't include health information about players
  * [x] *0x43 - party member removed*
  * [x] 0x45 - Transformed character in scope (transformation ring feature)
  * [x] *0x50 - guild join request*
  * [x] *0x51 - guild join response*
  * [x] 0x52 - guild list
  * [x] *0x53 - guild member kick result*
  * [x] *0x54 - guild master dialog*
  * [x] *0x55 - guild creation dialog*
  * [x] *0x56 - guild creation result*
  * [x] 0x5A - guild info
  * [x] 0x5B - guild user info (assign player to guild?)
  * [x] 0x5C - guild info (same as 5A, but just for one guild; in my test 5A does the job as well)
  * [x] 0x5D - guild deleted
  * [x] 0x60 - guild war request result
  * [x] 0x61 - guild war request
  * [x] 0x62 - guild war declared
  * [x] 0x63 - guild war ended
  * [x] 0x64 - guild score
  * [x] *0x81 - vault money set result*
  * [x] 0x86 - chaos mix result

In summary a manageable amount of messages changed:
  * guild messages are significantly different
  * all messages related to skills are different, as skill ids are 1 byte instead of 2 bytes (season 6)
  * magic effects are sent differently
    * The target id byte of the skill packet has a bit flag, if the effect of the skill was applied at it
  * items and appearance are significantly serialized differently as 0.75 doesn't support so many items and item options

**Comment by sven-n** [comment]
I managed to enter the game with 0.75 and Season 6 at the same time. I think it doesn’t make much sense, but it’s funny 😆 
The animation ids are different, so if you do something (e.g. rotating your player) in a season 6 client it looks buggy in 0.75 and vice-versa.
The old client is also very limited in terms of items - what should we do if the season 6 player is equipped with newer items or even of a master character class? And what to do with stuff the client doesn’t know yet like *Ability Gauge* and *Shield Bar*? Would need a lot of adaption logic.

So IMHO, it just makes sense to play 0.75 with a suitable GameConfiguration.

**Comment by fenos** [comment]
@sven-n i agree! I wouldn't bother about with items differences and skills and so on... As you said, who ever want to run a 0.75 they will have their own configuration and maybe share them with the community 😄 with the time

Regarding the rotation and basic animations i think they are important, we should be able to find a way to handle them

**Comment by fenos** [comment]
So far amazing job! i can't wait to try this out!

**Comment by sven-n** [comment]
I’ll start working on it soon again - had too many other things to do in the last 2-3 weeks :S
Anyway, I noticed an issue after walking and then attacking a monster. The character jumps back to where it initially spawned and started walking. Currently, no idea what’s wrong :(

**Comment by sven-n** [comment]
I made some improvements recently. The most obvious bugs should be solved by now :-)

One thing causes a bit of an headache to me... when the client closes a NPC dialog, it doesn't send any packet (in newer clients: C1 03 31). So after talking to an npc, the state of the player is staying in the "NPC Dialog Open" state on the server side. This prevents further gameplay, because this state is checked in a lot of actions :-(

**Comment by sven-n** [comment]
If somebody knows how to start a trade with MU in version 0.75, please tell me. With the /trade chat command, hovering the mouse over the other player, it doesn't work.
